File: 3rd Semester/CS2001-Data Structure/Lab/SortingAlgorithm/heap_sort.cpp
#include<iostream>
using namespace std;


void __heapify(int a[], int n, int i){
    int largest = i;
    int left = 2*i +1;
    int right = 2*i + 2;

    if(left<n  && a[left] > a[largest]){
        largest = left;
    }      
    if(right< n && a[right] > a[largest]){
        largest = right;
    }      

    if(largest != i){
        swap(a[i], a[largest]);
        __heapify(a, n, largest);
    }
}   //      end __heapify()

void __heapSort(int a[], int n){
    for (int i= n/2-1; i >=0; i--)
    {
        __heapify(a, n, i);
    }

    
    for (int i = n-1; i >= 0; i--)
    {
        swap(a[i], a[0]);
        __heapify(a, i, 0);
    }
    
    for (int i = 0; i < n; i++)
    {
        cout << "\n\t " << a[i];
    }
    
}

int main(){
    int a[] = {15, 5, 20, 1, 17, 10, 30};
    int n = sizeof(a)/sizeof(a[0]);

    __heapSort(a,n);

    return 0;
}

File: 3rd Semester/CS2001-Data Structure/Lab/SortingAlgorithm/heap_sort.py
def __heapify(lst, n, root):
    
    largest = root
    left = 2 * root +1
    right = 2 * root +2
    
    if left < n and lst[left] > lst[largest]:
        largest = left;
        
    if right < n and lst[right] > lst[largest]:
        largest = right
        
    if largest != root:
        
        lst[root], lst[largest] = lst[largest], lst[root]
        
        __heapify(lst, n, largest)

def __buildHeap(lst):
    
    n = len(lst)
    
    
    for i in reversed( range(n // 2)):
        
        __heapify(lst, n, i)
        
heap = [15, 5, 20, 1, 17, 10, 20]
print(heap)
__buildHeap(heap)
print(heap)


def __heapSort(lst):
    
    n = len(lst)
    
    __buildHeap(heap)
    
    for i in reversed( range(n)):
        
        lst[i], lst[0] = lst[0], lst[i]
        
        __heapify(lst, i, 0)
        
heap = [15, 5, 20, 1, 17, 10, 20]

print(heap)

__heapSort(heap)
print(heap)